From 07264bf37b527415b4bcdac9b9b44df9598c68fb Mon Sep 17 00:00:00 2001
From: Igor Grinberg <igrinberg@habana.ai>
Date: Mon, 09 Aug 2021 10:48:16 +0300
Subject: [PATCH] [SW-52119] gaudi: prevent tpc3 pcie access

The TPC3 pre-fetcher fails to access the PCIe address space in a way
that craches the PCIe bus.
To prevent this behaviour, block the TPC3 pre-fetcher from accessing the
PCIe address space by writing the high address bytes with 0x7F.
Additionally, block the high address register with protection bits to
prevent further changes from the host.

Change-Id: I5fdf5be471a0f138a24a3d3ea6b62be3d2a62fba
Signed-off-by: Igor Grinberg <igrinberg@habana.ai>
---

diff --git a/app/preboot/preboot.c b/app/preboot/preboot.c
index 9026443..8088580 100644
--- a/app/preboot/preboot.c
+++ b/app/preboot/preboot.c
@@ -369,6 +369,7 @@
 __weak void preboot_e2e_credits_init(void) {}
 __weak void preboot_device_id_init(void) {}
 __weak u32 preboot_device_id_get(void) { return 0; }
+__weak void preboot_engines_init(void) {}
 
 void preboot_main(void)
 {
@@ -428,6 +429,8 @@
 	if (hl_is_watchdog_reset())
 		i2csi_set_wd_err(HL_FW_RESET_CAUSE_WATCHDOG);
 
+	preboot_engines_init();
+
 	module_handle_cmds(cmd_ops, ARRAY_SIZE(cmd_ops));
 }
 
diff --git a/arch/gaudi/arm/asic.c b/arch/gaudi/arm/asic.c
index 40f59c8..51f4cda 100644
--- a/arch/gaudi/arm/asic.c
+++ b/arch/gaudi/arm/asic.c
@@ -12,6 +12,7 @@
 #include <dma1_qm_regs.h>
 #include <mme0_qm_regs.h>
 #include <mme1_qm_regs.h>
+#include <tpc3_cfg_regs.h>
 #include <tpc0_qm_regs.h>
 #include <tpc1_qm_regs.h>
 #include <cpu_if_regs.h>
@@ -760,3 +761,16 @@
 
 	cpu_boot_dev_sts_clrset(CPU_BOOT_DEV_STS0_E2E_CRED_EN, true);
 }
+
+#define TPC_PREFETCH_HMASK	0x7F
+void preboot_engines_init(void)
+{
+	/*
+	 * TPC3 pre-fetcher does not work for PCIe address space range.
+	 * Mask the high part of the pre-fetch address, so it will not get to
+	 * PCIe address space range.
+	 */
+	if (preboot_security_is_enabled())
+		regmap_writel(TPC_PREFETCH_HMASK,
+			      mmTPC3_CFG_ICACHE_BASE_ADDERESS_HIGH);
+}
diff --git a/arch/gaudi/arm/security.c b/arch/gaudi/arm/security.c
index a4c998f..080363a 100644
--- a/arch/gaudi/arm/security.c
+++ b/arch/gaudi/arm/security.c
@@ -1074,6 +1074,13 @@
 	/* offsets 0x900 - 0x980 */
 	prot_base = base + SNGL_REG_PROT_SIZE * 18;
 	mask = PROT_BIT(mmTPC0_CFG_PROT);
+	if (base == (mmTPC3_CFG_BASE & ~HL_CFG_BASE)) {
+		/*
+		 * TPC3 pre-fetcher does not work for PCIe address space range.
+		 * Block the register from host changes.
+		 */
+		mask &= PROT_BIT(mmTPC0_CFG_ICACHE_BASE_ADDERESS_HIGH);
+	}
 	regmap_writel(mask, PRIV_PROT_REG(prot_base));
 
 	/* offsets 0x980 - 0xA00 */
diff --git a/include/preboot.h b/include/preboot.h
index 7f797fc..7acc655 100644
--- a/include/preboot.h
+++ b/include/preboot.h
@@ -214,6 +214,7 @@
 void preboot_init_pcie_phy_fw_update(void);
 void preboot_disable_pcie_ltr(void);
 void preboot_init_pcie_phy(void);
+void preboot_engines_init(void);
 
 #endif /* __ASSEMBLY__ */
 
